pgmax/groups/variables.py

```diff
@@ -106,29 +176,57 @@ class VariableDict(groups.VariableGroup):
 
     Args:
         num_states: The size of the variables in this variable group
-        variable_names: A tuple of all names of the variables in this variable group
-
+        variable_names: A tuple of all names of the variables in this variable group.
+            Note that we overwrite variable_names to add the hash of the VariableDict
     """
 
-    num_states: int
     variable_names: Tuple[Any, ...]
+    num_states: int
+
+    def __post_init__(self):
+        num_states = np.full(
+            (len(self.variable_names),), fill_value=self.num_states, dtype=np.int32
+        )
+        object.__setattr__(self, "num_states", num_states)
 
-    def _get_names_to_variables(self) -> OrderedDict[Tuple[int, ...], nodes.Variable]:
-        """Function that generates a dictionary mapping names to variables.
+        # Only compute the hash once
+        object.__setattr__(self, "this_hash", id(self))
+
+        hash_and_names = tuple(
+            (self.__hash__(), var_name) for var_name in self.variable_names
+        )
+        object.__setattr__(self, "variable_names", hash_and_names)
```

Comment:
Make this into a cached property

Reply:
no: we are overwritting `variable_names` to add the hash

Reply:
OK I missed that...That naming is confusing, though. Can we change the `variable_names` containing the hash to `variable_identifiers`? Those are not really names in the first place